Added support for bc6h format

theKlaxon opened the pull request:
- Added bc6h compression function.
- Added bc6h to all enums + lists.
- Updated Compressonator to 4.5.52 for proper bc6h support.

theKlaxon committed:
Added support for bc6h format

- Added bc6h compression function.
- Added bc6h to all enums + lists.
- Updated Compressonator to 4.5.52 for proper bc6h support.

craftablescience commented:
The prebuilt static library for Linux will need to be updated as well, I have builds of the latest version at https://github.com/craftablescience/compressonator-compiler/releases/tag/v4.5-f4b53d7-2 if you need it

theKlaxon commented:
You got it, I'll make the change!

SCell555 commented:
Just to note, IMAGE_FORMAT_BC6H is signed BC6H.

JJL772 commented on VTFLib/VTFFormat.h:
```diff
@@ -95,6 +95,7 @@ typedef enum tagVTFImageFormat
 	IMAGE_FORMAT_ATI2N,						//!<  = Red, Green BC5 compressed format - 8 bpp
 	IMAGE_FORMAT_ATI1N,						//!<  = Red BC4 compressed format - 4 bpp
 
+	IMAGE_FORMAT_BC6H = 69,					//!<  = Red, Green, Blue, BC6H compressed format - 8 bpp
```
`IMAGE_FORMAT_BC6H` should be 71

theKlaxon replied:
Cool, I'll change the rest of the enums / lists to reflect that also.

JJL772 commented on VTFLib/VTFFile.cpp:
```diff
@@ -3556,7 +3629,7 @@ static SVTFImageConvertInfo VTFImageConvertInfo[IMAGE_FORMAT_COUNT] =
 	{},
 	{},
-	{},
+	{	  8,  0,  0,  0,  0,  0,	-1, -1, -1, -1,	 vlTrue, vlTrue,	NULL, NULL,			IMAGE_FORMAT_BC6H},
```
This will need to get moved to after `IMAGE_FORMAT_BC7` when you update the enum (see the later comment)

theKlaxon replied:
✔

JJL772 commented on VTFLib/VTFFile.cpp:
```diff
@@ -3031,7 +3032,7 @@ static SVTFImageFormatInfo VTFImageFormatInfo[] =
 	{},
 	{},
-	{},
+	{ "BC6H",					8,  0,  0,  0,  0,  0, vlTrue,  vlTrue  },			// IMAGE_FORMAT_BC6H
```
This will need to get moved to after IMAGE_FORMAT_BC7 when you update the enum (see the later comment)

theKlaxon replied:
✔

JJL772 commented on VTFLib/VTFFile.cpp:
```diff
@@ -3350,6 +3355,74 @@ vlBool CVTFFile::CompressBCn(const vlByte *lpSource, vlByte *lpDest, vlUInt uiWi
+	srcTexture.dwWidth = uiWidth;
+	srcTexture.dwHeight = uiHeight;
+	srcTexture.dwPitch = 4 * uiWidth;
+	srcTexture.format = CMP_FORMAT_RGBA_8888;
```
Is the input to this function always assumed to be RGBA8888?

theKlaxon replied:
ah, it was while i was testing things out. ill go ahead and add the SourceFormat param back to make sure an edge case isn't hit.

theKlaxon replied:
✔

JJL772 replied:
My issue with this code is less about assuming `lpSource` is always RGBA8888 and more about the potential data loss. If textures are getting squished down to RGBA8888 before being turned into BC6H, what's the point of using BC6H at all

craftablescience replied:
This is still blocking merge, basically what JJ said

JJL772 reviewed (changes requested):
Thanks for this work! I have a couple small comments/questions.

theKlaxon commented:
> Just to note, IMAGE_FORMAT_BC6H is signed BC6H.

Would you like me to note it at the end of the format as IMAGE_FORMAT_BC6H_SF ?

SCell555 commented:
No, since there won't be an unsigned version. And we don't need to be weird like compressonator with format naming.

theKlaxon committed:
Updated win and linux to latest compressonator

- Moved BC6H format to 71
- Fixed incorrect dwDataSize given to midTexture in VTFFile.cpp

theKlaxon committed:
Added necessary linux libs and tested compilation

- Updated CMakeLists.txt to reflect new dependencies.

theKlaxon committed:
Fixed small linux bug with RGBA_16 dwPitch

craftablescience commented on CMakeLists.txt:
```diff
@@ -61,21 +61,35 @@ endif()
-		target_link_directories(vtflib_static PUBLIC thirdparty/lib/x64)
-		target_link_libraries(vtflib_static PUBLIC CMP_Compressonator pthread)
+		target_link_directories(vtflib_static PUBLIC thirdparty/lib/linux_x86_64)
+		target_link_libraries(vtflib_static 
+		PUBLIC CMP_Compressonator pthread
+		PUBLIC CMP_Core pthread
+		PUBLIC CMP_Core_AVX pthread
+		PUBLIC CMP_Core_AVX512 pthread
+		PUBLIC CMP_Core_SSE pthread
+		PUBLIC CMP_Framework pthread
+		)
```
```suggestion
		target_link_libraries(vtflib_static PUBLIC
				CMP_Compressonator
				CMP_Core
				CMP_Core_AVX
				CMP_Core_AVX512
				CMP_Core_SSE
				CMP_Framework
				pthread
		)
```

craftablescience commented on CMakeLists.txt:
```diff
@@ -61,21 +61,35 @@ endif()
-		target_link_directories(vtflib PUBLIC thirdparty/lib/x64)
-		target_link_libraries(vtflib PUBLIC CMP_Compressonator pthread)
+		target_link_directories(vtflib PUBLIC thirdparty/lib/linux_x86_64)
+		target_link_libraries(vtflib 
+		PUBLIC CMP_Compressonator pthread
+		PUBLIC CMP_Core pthread
+		PUBLIC CMP_Core_AVX pthread
+		PUBLIC CMP_Core_AVX512 pthread
+		PUBLIC CMP_Core_SSE pthread
+		PUBLIC CMP_Framework pthread
+		)
```
```suggestion
		target_link_libraries(vtflib PUBLIC
				CMP_Compressonator
				CMP_Core
				CMP_Core_AVX
				CMP_Core_AVX512
				CMP_Core_SSE
				CMP_Framework
				pthread
		)
```

theKlaxon committed:
Removed redundant 'public' tag in CMakeLists.txt

Co-authored-by: craftablescience <lauralewisdev@gmail.com>

theKlaxon committed:
Removed redundant 'public' tag in CMakeLists.txt

Co-authored-by: craftablescience <lauralewisdev@gmail.com>

theKlaxon committed:
Updated to Craftablescience's linux libs

theKlaxon committed:
Removed CMP_Framework

craftablescience commented on CMakeLists.txt:
```diff
@@ -61,21 +61,35 @@ endif()
+				CMP_Core_AVX
+				CMP_Core_AVX512
+				CMP_Core_SSE
+				pthread
```
use cmake's `Threads::Threads` target instead of `pthread`

from stackoverflow:
```cmake
set(CMAKE_THREAD_PREFER_PTHREAD ON)
set(THREADS_PREFER_PTHREAD_FLAG ON)
find_package(Threads REQUIRED)

add_executable(test test.cpp)
target_link_libraries(test Threads::Threads)
```

also, is this actually required? i cant remember any vtflib code that uses pthreads or any threading stuff in `std`

craftablescience commented on CMakeLists.txt:
```diff
@@ -61,21 +61,35 @@ endif()
+				CMP_Core_AVX
+				CMP_Core_AVX512
+				CMP_Core_SSE
+				pthread
```
ditto

craftablescience commented on VTFLib/VTFFile.cpp:
```diff
@@ -440,6 +440,7 @@ static CMP_FORMAT GetCMPFormat( VTFImageFormat imageFormat, bool bDXT5GA )
 	// Swizzle is technically wrong for below but we reverse it in the shader!
 	case IMAGE_FORMAT_ATI2N:			return CMP_FORMAT_ATI2N;
 
+	case IMAGE_FORMAT_BC6H:				return CMP_FORMAT_BC6H;
```
nit: should come after BC7

craftablescience commented on VTFLib/VTFFile.cpp:
```diff
@@ -3066,6 +3068,7 @@ vlUInt CVTFFile::ComputeImageSize(vlUInt uiWidth, vlUInt uiHeight, vlUInt uiDept
 	case IMAGE_FORMAT_DXT3:
 	case IMAGE_FORMAT_DXT5:
 	case IMAGE_FORMAT_ATI2N:
+	case IMAGE_FORMAT_BC6H:
```
nit: should be after BC7

craftablescience commented on VTFLib/VTFFile.cpp:
```diff
@@ -3857,6 +3932,7 @@ vlBool CVTFFile::Convert(const vlByte *lpSource, vlByte *lpDest, vlUInt uiWidth,
 		case IMAGE_FORMAT_DXT5:
 		case IMAGE_FORMAT_ATI2N:
 		case IMAGE_FORMAT_ATI1N:
+		case IMAGE_FORMAT_BC6H:
```
nit: should be below.. yeah you get the idea

craftablescience commented on VTFLib/VTFFormat.h:
```diff
@@ -94,8 +94,9 @@ typedef enum tagVTFImageFormat
 	IMAGE_FORMAT_ATI2N,						//!<  = Red, Green BC5 compressed format - 8 bpp
 	IMAGE_FORMAT_ATI1N,						//!<  = Red BC4 compressed format - 4 bpp
-
+	
```
nit: extra space

craftablescience reviewed (changes requested):
I mostly just have a few nitpicks, once JJ's comment and pthread stuff gets resolved this is good for merge
